Formula/cp2k.rb

```diff
@@ -48,35 +57,95 @@ def install
       s.gsub!(/DFLAGS *=/, "DFLAGS = -D__FFTW3")
       s.gsub!(/FCFLAGS *=/, "FCFLAGS = -I#{Formula["fftw"].opt_include}")
       s.gsub!(/LIBS *=/, "LIBS = #{libs.join(" ")}")
     end
 
     # MPI versions link to scalapack
-    inreplace Dir["arch/Darwin-gfortran.p*"],
+    inreplace Dir["arch/#{arch}-gfortran.p*"],
               /LIBS *=/, "LIBS = -L#{Formula["scalapack"].opt_prefix}/lib"
 
     # OpenMP versions link to specific fftw3 library
-    inreplace Dir["arch/Darwin-gfortran.*smp"],
+    inreplace Dir["arch/#{arch}-gfortran.*smp"],
               "-lfftw3", "-lfftw3 -lfftw3_threads"
 
+    # Remove flags for unused dependencies on Linux.
+    # The build system has been completely refactored upstream and
+    # these manual fixes will not be needed in the next release.
+    unless OS.mac?
+      inreplace Dir["arch/#{arch}-gfortran.*smp"] do |s|
+        s.gsub!("-D__LIBVORI", "")
+        s.gsub!("-D__LIBXSMM", "")
+        s.gsub!("-D__SPGLIB", "")
+
+        s.gsub!("-I$(LIBXSMM_INC)", "")
+
+        s.gsub!("$(LIBVORI_LIB)/libvori.a", "")
+        s.gsub!("$(SPGLIB_LIB)/libsymspg.a", "")
+        s.gsub!("$(LIBXSMM_LIB)/libxsmmf.a", "")
+        s.gsub!("$(LIBXSMM_LIB)/libxsmm.a", "")
+        s.gsub!("$(LIBPATH)/liblapack.a", "")
+
+        s.gsub!("$(GNU_PATH)/fftw/3.3.9/include", Formula["fftw"].opt_include)
+        s.gsub!("$(GNU_PATH)/fftw/3.3.9/lib", Formula["fftw"].opt_lib)
+        s.gsub!("$(GNU_PATH)/libint/2.6.0-lmax-6", libexec)
+        s.gsub!("$(GNU_PATH)/libxc/5.1.4/include", Formula["libxc"].opt_include)
+        s.gsub!("$(GNU_PATH)/libxc/5.1.4/lib", Formula["libxc"].opt_lib)
+      end
+
+      inreplace "arch/#{arch}-gfortran.ssmp" do |s|
+        s.gsub!("$(LIBPATH)/libblas.a", "#{Formula["openblas"].opt_lib}/libopenblas.a")
+        s.gsub!("-static", "")
+      end
+
+      inreplace "arch/#{arch}-gfortran.psmp" do |s|
+        s.gsub!("include       $(MPI_PATH)/plumed2/2.6.2/lib/plumed/src/lib/Plumed.inc.static", "")
+        s.gsub!("-D__ELPA", "")
+        s.gsub!("-D__PLUMED2", "")
+        s.gsub!("-D__SIRIUS", "")
+        s.gsub!("-D__SCALAPACK", "")
+
+        s.gsub!("-I$(ELPA_INC)/elpa -I$(ELPA_INC)/modules", "")
+        s.gsub!("-I$(SIRIUS_INC)", "")
+        s.gsub!("-I$(SPGLIB_INC)", "")
+
+        s.gsub!("$(PLUMED_DEPENDENCIES)", "")
+        s.gsub!("$(ELPA_LIB)/libelpa_openmp.a", "")
+        s.gsub!("${SIRIUS_LIB}/libsirius.a", "")
+        s.gsub!("$(GNU_PATH)/SpFFT/0.9.13/lib/libspfft.a", "")
+        s.gsub!("$(GNU_PATH)/SpLA/1.2.1/lib/libspla.a", "")
+        s.gsub!("$(GNU_PATH)/hdf5/1.12.0/lib/libhdf5.a", "#{Formula["hdf5"].opt_lib}/libhdf5.a")
+        s.gsub!("$(GNU_PATH)/OpenBLAS/0.3.15/lib/libopenblas.a",
+                "#{Formula["openblas"].opt_lib}/libopenblas.a")
+        s.gsub!("$(LIBPATH)/libz.a", "#{Formula["zlib"].opt_lib}/libz.a")
+        s.gsub!("$(MPI_LIBRARY_PATH)/libscalapack.a", "")
+      end
+    end
```

Comment:
This looks complicated and fragile. We need a simpler and more maintainable way of doing this, I think.

Reply:
I don't know if the comment made it clear, but upstream has completely refactored the build system and this will not be needed at the next release.  The new Linux build file doesn't hard code paths to dependencies anymore, so we'll just be able to pass the the proper paths in as arguments without needing to patch anything: https://github.com/cp2k/cp2k/blob/master/arch/Linux-gnu-x86_64.psmp.  What we're doing is just a temporary solution until the next release and it should be much cleaner and simpler at that point.

Reply:
Let's wait until the next release then. I'm a bit wary of carrying around dozens of `inreplace` lines in a formula.

Reply:
Hmm, it does feel a bit weird to me to leave a formula in a broken state when we know how to fix it, even the fix is quite tedious.  Would it be more manageable if I used a patch to delete the unused lines?  That would get us down to just a handful of `inreplace` statements for the cases where are pointing it to brewed dependencies.

Reply:
A patch is basically the same. Both add a lot of technical debt to the formula, and this one is already particularly [painful to update](https://github.com/Homebrew/homebrew-core/pulls?q=is%3Apr+is%3Aclosed+in%3Atitle+cp2k+status%3Afailure). I don't want us to get stuck in a position where we're unable to update the formula because a complicated patch needs reworking.

Reply:
I see that it’s given us issues in the past.  I guess the concern here is that even if a future update does let us drop the patching/inreplace, we’d be creating more work for ourselves by having to clean all this up for that update.  

If nothing else this will provide a useful starting point for when the next release does come out, but for now it does seem better to wait until the new release.

Reply:
Thanks @danielnachun. Apologies for having your work here go to waste.

If you think there's demand for this formula on Linux, it can be hosted in an external tap in the meantime.

Reply:
Yeah I think that would make sense in the event that we do get people asking.  From what I can tell there seem to be some other optional dependencies that could probably eventually be added to make this more useful in Linux HPC environments.  While we wait for upstream to make a new release I may look into these other dependencies to see how difficult they are to build and whether they may warrant their own formula or should be considered resources.